### DISCUSSION TITLE
Add camera API for exchanging WebRTC offer/answer asynchronously

### DISCUSSION THREAD

**Discussion by MartinHjelmare** [home-assistant/architecture#1165] [Entity Models] [answered]
## Background

The main benefit of WebRTC is that it allows a direct connection between the device, in our case the camera, and the client, which is our frontend. Besides setting the correct supported feature and stream type, it currently requires the camera platform to implement `async_handle_web_rtc_offer`. This should pass the frontend's SDP offer to the device and return back the answer. Today this API only supports a synchronous communication where the frontend expects to get all candidates upfront. Since this exchange can be slow and take upwards a minute on some clients, the user experience is currently not good across the board, when setting up a WebRTC stream.

## Proposal

Add an asynchronous API where the exchange can happen piecewise. The supporting camera entity should implement the following methods:
- `async_handle_async_webrtc_offer`: To initialize a WebRTC stream. Any messages/errors coming in async should be forwared to the frontend with the `send_message` callback.
- `async_on_webrtc_candidate`: The frontend will call it with any candidate coming in after the offer is sent.
- `close_webrtc_session`: The frontend will call it when the stream is closed. Can be used to clean up things.

Default methods for these APIs are available on the base camera entity, but can be overridden by integrations.

```python
class MyCamera(Camera):

    async def async_handle_async_webrtc_offer(
        self, offer_sdp: str, session_id: str, send_message: WebRTCSendMessage
    ) -> None:
        """Handle the async WebRTC offer.

        Async means that it could take some time to process the offer and responses/message
        will be sent with the send_message callback.

        This method is used by cameras with CameraEntityFeature.STREAM and StreamType.WEB_RTC.

        An integration overriding this method must also implement async_on_webrtc_candidate.

        Integrations can override with a native WebRTC implementation.
        """

    async def async_on_webrtc_candidate(self, session_id: str, candidate: str) -> None:
        """Handle a WebRTC candidate."""

    @callback
    def close_webrtc_session(self, session_id: str) -> None:
        """Close a WebRTC session."""
```

An important difference interface-wise between the sync API and async API is that the candidates are not part of the offer in the async API. Explicit support for forwarding the candidates from the frontend will need to be implemented via `async_on_webrtc_candidate`. In the sync API the candidates are included in the initial offer.

## Consequences

- The existing sync API will keep working, via a compatibility layer, which asks the frontend to get candidates up front with the async API. A deprecation of the sync API can be a future discussion, if wanted. Existing integrations that use the sync API will keep working, and are encouraged to migrate to the async API as possible.
- A default async API implementation will be available for all integrations that configure the camera entity to support WebRTC and don't need a native implementation. See: https://github.com/home-assistant/core/pull/127981
- New integrations that need a native implementation should implement the async API.

**Comment by allenporter** [comment]
Thank you for starting the discussion.  Overall very supportive. This can solve the needs of sessions in https://github.com/home-assistant/architecture/discussions/1151 👍🏼  Agree on moving everything to the new API.

I am not certain that the proposal fully handles compatibility for ice candidates (the last two consequences 🟡). Below, I am not taking a strong opinion on solution, just want to articulate the problem.

Thoughts:
- Background: The frontend currently listens for ice candidates, blocks until the frontend client resolves them, then appends the candidates to the offer (https://github.com/home-assistant/frontend/pull/13942). The new behavior will fix this, as the current behavior is a workaround for lack of an async API. The current plan is the frontend will change to stop doing this to use the above proposed API.  (🟢)
- If an integration wants to use the sync behavior still, then it can append the candidates to the offer itself through the async candidate API (🟢)
- However, the async API does not provide any notion of when all ice candidates are have been received and different candidates may be used in different network scenarios. (🟡 )

My impression is the currently proposed default async implementation does not do anything about ice candidates at all, and would break async ice candidate handling. Options I can think of for default implementation or for integrations moving to the new API are either to wait for some candidates, or set a short timeout, or some hybrid approach, before sending the offer. 

Scope:  The existing integrations using the sync APIs are nest, frigate, and rtsp_to_webrtc. My current understanding is:
- Nest: Does not use ice candidates and does not need further consideration. It can move to new APIs immediately, or can use any default implementation that ignores ice candidates 🟢
- Frigate:  Uses `go2rtc` which does have a signaling API for async candidates.  It needs a solution while using the current sync API to preserve ice candidate behavior  🔴  but once change it can natively support async candidates 🟢
-`rtsp_to_webrtc` (via provider APIs) supports servers that do not have a signaling API for candidates and needs an approach 🔴 . 

There are no other known usages of these APIs.

**Reply by edenhaus** [reply]
We should update the proposal that the `sync` variant is not deprecated, and we currently have no specific plans for when and if we deprecate it. Nevertheless, integrations should use/migrate to the async one for better performance/user experience, if possible.

> My impression is the currently proposed default async implementation does not do anything about ice candidates at all, and would break async ice candidate handling. Options I can think of for default implementation or for integrations moving to the new API are either to wait for some candidates, or set a short timeout, or some hybrid approach, before sending the offer.

https://github.com/home-assistant/core/pull/127981 will not break anything with the sync approach, and we fall back automatically to the sync approach if async is not supported. The PR mentioned will add a boolean attribute named `get_candidates_upfront` to the client WebRTC config, which tells the frontend whether it needs to get all candidates before sending the offer.

Please keep in mind that the difference between the sync and async approaches is that the offer in the async approach can have candidates but doesn't need to. Most candidates will be received after the offer is sent. In comparison to the sync one, where all candidates are included in the offer.

> - Frigate: Uses go2rtc which does have a signaling API for async candidates. It needs a solution while using the current sync API to preserve ice candidate behavior 🔴 but once change it can natively support async candidates 🟢
> - rtsp_to_webrtc (via provider APIs) supports servers that do not have a signaling API for candidates and needs an approach 🔴 .

As we have full backward compatibility with the sync one, in my opinion, both red points are solved, as the integrations continue to work as they currently do. Both will use the sync approach and the new async approach can be added to the integrations in further PRs if possible

**Reply by MartinHjelmare** [reply]
I've updated the proposal.

**Reply by allenporter** [reply]
Looks good, approved on my end.

Having the option to preserve the behavior of getting candidates up front is sufficient to preserve the existing behavior for the shim.

Slight nit in wording: I would just add that we can use the mechanism to get candidates up front with the async API such that there is a migration path forward to drop usage of the sync API (with the caveats that it performs worse, so it is not desired to be used, just for migration e.g. rtsp_to_webrtc when configured with a stun server).

**Reply by MartinHjelmare** [reply]
I've clarified what the compatibility layer between sync and async APIs means.

**Comment by edenhaus** [comment, marked as answer]
Approved on my end and also by @allenporter in https://github.com/home-assistant/architecture/discussions/1165#discussioncomment-11075442
